### ISSUE TITLE
function that returns union with `void` can be incorrectly narrowed

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#46837]
# Bug Report

### 🔎 Search Terms

void narrow

### 🕗 Version & Regression Information

4.6.0-20211117

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.0-dev.20211117#code/MYewdgzgLgBAhgLhgCgG5wDYFcCmSxYC2ARjgE4CUMAvAHwyogCWAJjSk1XTEwLABQoSLGJI0mXPiKlKNeo1YwAPjGhkmYAObs4AgUOgN2xZAEYKApgDMUqKgG8BMZwwB0UEAFUADt-IBhOAgcZCoAejCyLDAoJkIcGHIyEDIkVHcvXwCghKYIGDAQWDgYK2jgWPAYAQBfIA)

### 💻 Code

```ts
const a: (value: number) => void = (i) => i
const b: (value: number) => void | string = a

const v = b(1)
if (v) {
    v.toUpperCase() //runtime error: v.toUpperCase is not a function 
}
```

### 🙁 Actual behavior
no error

### 🙂 Expected behavior
value doesn't get narrowed. perhaps `void` should not be falsy because you can't be sure its value will be falsy at runtime

**Comment by MartinJohns** [comment]
Related: #42709

**Comment by RyanCavanaugh** [comment]
I think we should do some patchwork and just disallow writing `void | T` in `--strict`. It's never what you want.

**Comment by KotlinIsland** [comment]
@RyanCavanaugh based

**Comment by mdbetancourt** [comment]
> I think we should do some patchwork and just disallow writing `void | T` in `--strict`. It's never what you want.

i don't think so void !== undefined
```ts
function convert<T>(t: T): void | T { // ok
  if(typeof t === 'string') {
    return t
  }
}
function convert2<T>(t: T): undefined | T { // error here "A function whose declared type is neither 'void' nor 'any' must return a value"
  if(typeof t === 'string') {
    return t
  }
}
```

**Comment by RyanCavanaugh** [comment]
@mdbetancourt `noImplicitReturns` means no implicit returns. If you don't like the error in the second case, you should turn that option off.

**Comment by mdbetancourt** [comment]
what about this case extending his case

```ts
const a = (i: number[]) => i.forEach(console.log)

const b: (value?: number) => void | false = (v) => v ? a([v]) : false
                                               ^ undefined or anything else is not gonna work

```

**Comment by RyanCavanaugh** [comment]
`unknown` is equally correct there. We could change `forEach` to return `undefined` and then it can be `undefined | false`.

**Comment by mdbetancourt** [comment]
but that could break some codes isn't? e.g:

```ts
function voidFn() {}

const log = [].forEach as () => undefined
const h: typeof log = voidFn
//    ^ Type 'void' is not assignable to type 'undefined'
```

a real case could be
```ts
function altLogger(text: string) {}

// if you change forEach signature you have to change every void signature then
const log = console.log as (text: string) => undefined
const h: typeof log = altLogger
//    ^ Type 'void' is not assignable to type 'undefined'
```

**Comment by marcospgp** [comment]
> what about this case extending his case
> 
> ```ts
> const a = (i: number[]) => i.forEach(console.log)
> 
> const b: (value?: number) => void | false = (v) => v ? a([v]) : false
> //                             ^ undefined or anything else is not gonna work
> ```

I think this should error, and a function should not be able to return a union that includes void. It either returns something or not at all.

**Comment by robbiespeed** [comment]
Here's another example of the issue, this time without direct use of `void | T`:
```ts
function run (fn: (() => void) | (() => number)): number {
  const v = fn(); // results in void | number
  // void behaving like undefined
  if (typeof v === "object") {
    v; // never
    return -1;
  }
  if (v !== undefined) {
    return v; // TS thinks it's a number (but could really be anything)
  }

  return -1;
}

const foo = () => "hello";

const num: number = run(foo); // allowed because void allows any return value but ends up returning a string that thinks it's a number

typeof num === "string"; 
```
[playground](https://www.typescriptlang.org/play/?ts=5.5.4#code/GYVwdgxgLglg9mABAJ3IgFMMAuD6CUiAvAHyIBucMAJoQD56GmJggC2ARgKbL766tOPRAG8AUIkQQEAZygViiLAQDciAPTqUXGSAA2UGYhhJKNRA0HdkEjVrPVE3ABYBDciYDmiPTADWXIjg1FzAJlzUtjDAGFAAngAOXHAx5MRERIgARHAcAFZc0FmE4pKS5GqaLFzkPLaSyFxQIMhIALQAjCq2AL5RMehpAIQZQWAhYWARJfXaza0UlVoAKgDKiFDOJn5GMFAA5EauLOzWGBwg8tL6jo2uenpxToGuYHGbXvi9YraN8+1dMR9MTSMByJRwOCKAjEMhZZxcB5wLLdEGyeSCASnYSZVBgTCQ-BLRD3PRwADuEWeEFcIBkgQcJKR5KObzmLVM9xAgQu8i44yMIAS7NaXhJiDkyDFm1c8g+YB2xgORxOQhsYniSRSqvSmSykq8KMQYiAA)

> I think we should do some patchwork and just disallow writing void | T in --strict. It's never what you want.

@RyanCavanaugh I don't think that will cut it. In the example above you don't need to write it, TS will infer it. I suppose you could disallow any union containing void somewhere no matter how deep, that seems complicated though and might be overly limiting. What about any time TS encounters `void | T` either explicitly or inferred it converts it to `unknown`?

**Comment by RyanCavanaugh** [comment]
I agree with @robbiespeed 's analysis. Really the only "good" fix here is #42709, so let's just track this there.
